internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,58 +116,331 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+		}
+	}
+
+	// Should not happen due to CRD defaulting
+	if gitHubAPIConfig.Host == nil || len(*gitHubAPIConfig.Host) < 1 {
+		return buildInvalidHost("", "must not be empty"), nil
+	}
+
+	host := *gitHubAPIConfig.Host
+	hostPort, addressParseErr := endpointaddr.Parse(host, 443)
+	if addressParseErr != nil {
+		// addressParseErr is not recoverable. It requires user interaction, so do not return the error.
+		return buildInvalidHost(host, addressParseErr.Error()), nil
+	}
+
+	return &metav1.Condition{
+		Type:    HostValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is valid", host),
+	}, &hostPort
+}
+
+func (c *gitHubWatcherController) validateTLSConfiguration(tlsSpec *v1alpha1.TLSSpec) (*metav1.Condition, *x509.CertPool) {
+	certPool, _, buildCertPoolErr := pinnipedcontroller.BuildCertPoolIDP(tlsSpec)
+	if buildCertPoolErr != nil {
+		// buildCertPoolErr is not recoverable with a resync.
+		// It requires user interaction, so do not return the error.
+		return &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidTLSConfig",
+			Message: fmt.Sprintf("spec.githubAPI.tls.certificateAuthorityData is not valid: %s", buildCertPoolErr),
+		}, nil
+	}
+
+	return &metav1.Condition{
+		Type:    TLSConfigurationValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: "spec.githubAPI.tls.certificateAuthorityData is valid",
+	}, certPool
+}
+
+func (c *gitHubWatcherController) validateGitHubConnection(
+	hostPort *endpointaddr.HostPort,
+	certPool *x509.CertPool,
+	validSoFar bool,
+) (*metav1.Condition, string, *http.Client, error) {
+	if !validSoFar {
+		return &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: "unable to validate; see other conditions for details",
+		}, "", nil, nil
+	}
+
+	conn, tlsDialErr := tls.Dial("tcp", hostPort.Endpoint(), ptls.Default(certPool))
+	if tlsDialErr != nil {
+		return &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "UnableToDialServer",
+			Message: fmt.Sprintf("cannot dial server spec.githubAPI.host (%q): %s", hostPort.Endpoint(), buildDialErrorMessage(tlsDialErr)),
+		}, "", nil, tlsDialErr
+	}
+
+	return &metav1.Condition{
+		Type:    GitHubConnectionValid,
+		Status:  metav1.ConditionTrue,
+		Reason:  upstreamwatchers.ReasonSuccess,
+		Message: fmt.Sprintf("spec.githubAPI.host (%q) is reachable and TLS verification succeeds", hostPort.Endpoint()),
+	}, fmt.Sprintf("https://%s", hostPort.Endpoint()), c.httpClientBuilder(certPool), conn.Close()
+}
+
+// buildDialErrorMessage standardizes DNS error messages that appear differently on different platforms, so that tests and log grepping is uniform.
+func buildDialErrorMessage(tlsDialErr error) string {
+	reason := tlsDialErr.Error()
```

Comment:
Pondering if there is any case were this error message might be excessively long and we would want to use this helper:

```golang
pinnipedcontroller.TruncateMostLongErr(err))
```

I recall in JWTAuthenticator the 404 error case might contain an HTML page in the output, which could be quite a bit of data.

Reply:
But the JWTAuthenticator does an actual HTTP request, not just a Dial, so I don't know that this is likely to get much more than a line or two of an error message.